pkg/controller/servicemesh/controlplane/hooks.go

```diff
@@ -69,146 +58,111 @@ func (r *ControlPlaneReconciler) processNewObject(object *unstructured.Unstructu
 }
 
 func (r *ControlPlaneReconciler) processDeletedObject(object *unstructured.Unstructured) error {
-	switch object.GetKind() {
-	case "Route":
-		if object.GetName() == "kiali" {
-			return r.processDeletedKialiRoute(object)
-		}
-	}
 	return nil
 }
 
-var (
-	grafanaRegexp = regexp.MustCompile("(?s)(grafana:.*?url:).*?\n")
-	jaegerRegexp  = regexp.MustCompile("(?s)(tracing:.*?url:).*?\n")
-)
-
 func (r *ControlPlaneReconciler) patchKialiConfig(object *unstructured.Unstructured) error {
-	r.Log.Info("patching kiali ConfigMap", object.GetKind(), object.GetName())
-	configYaml, found, err := unstructured.NestedString(object.UnstructuredContent(), "data", "config.yaml")
-	if err != nil {
-		// This shouldn't occur if it's really a ConfigMap, but...
-		r.Log.Error(err, "could not parse kiali ConfigMap")
-		return err
-	} else if !found {
-		r.Log.Error(nil, "failed to patch kiali ConfigMap")
-		return nil
+	r.Log.Info("patching kiali CR", object.GetKind(), object.GetName())
+
+	// get jaeger URL and enabled flags from Kiali CR
+	jaegerURL, found, err := unstructured.NestedString(object.UnstructuredContent(), "spec", "external_services", "tracing", "url")
+	if !found || err != nil {
+		jaegerURL = ""
+	}
+	jaegerEnabled, found, err := unstructured.NestedBool(object.UnstructuredContent(), "spec", "external_services", "tracing", "enabled")
+	if !found || err != nil {
+		jaegerEnabled = true // if not set, assume we want it - we'll turn this off if we can't auto-detect
+	}
+
+	// if the user has not yet configured this, let's try to auto-detect it now.
+	if len(jaegerURL) == 0 && jaegerEnabled {
+		r.Log.Info("attempting to auto-detect jaeger for kiali")
+		jaegerRoute := &unstructured.Unstructured{}
+		jaegerRoute.SetAPIVersion("route.openshift.io/v1")
+		jaegerRoute.SetKind("Route")
+		// jaeger is the name of the Jaeger CR
+		err = r.Client.Get(context.TODO(), client.ObjectKey{Name: "jaeger", Namespace: object.GetNamespace()}, jaegerRoute)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				r.Log.Error(err, "error retrieving jaeger route - will disable it in Kiali")
+				// we aren't going to return here - Jaeger is optional for Kiali; Kiali can still run without it
+			}
+			jaegerEnabled = false
+		} else {
+			jaegerURL, _, _ = unstructured.NestedString(jaegerRoute.UnstructuredContent(), "spec", "host")
```

Comment:
This may be my misunderstanding of this unstructured API - where is this extra call to the API server?

I thought `jaegerRoute.UnstructuredContent()` merely obtains the json that was retrieved already from the r.Client.Get() call from above. Are you saying `unstructured.NestedString` or `jaegerRoute.UnstructuredContent()` makes another call to the API server?

I see only two calls to the API server - one to get jaeger route, one to get grafana route.

Reply:
Never mind.  I don't know what I was seeing before.  This is fine as is.